Extract common code from AbstractJvmQuarkusTestExtension and QuarkusMainTestExtension to AppMakerHelper

aloubyansky opened the pull request:
This is the first step towards reducing duplicated code in a couple of test extension classes.

aloubyansky committed:
Extract common code from AbstractJvmQuarkusTestExtension and
QuarkusMainTestExtension to AppMakerHelper

holly-cummins commented on test-framework/junit5/src/main/java/io/quarkus/test/junit/AppMakerHelper.java:
```diff
@@ -86,65 +75,75 @@ private PrepareResult createAugmentor(final Class<?> requiredTestClass, String d
+        final QuarkusTestProfile profileInstance = new ClassCoercingTestProfile(profile.getConstructor()
+                .newInstance());
+        final Map<String, String> additional = new HashMap<>();
+        // TODO we make this twice, also in abstractjvmextension can we streamline that?
```
Does this comment still apply? I think you've just fixed that problem, and can delete the TODO.

holly-cummins replied:
... or maybe the duplicated code has been fixed, but the execution still happens twice?

aloubyansky replied:
The test profile is now created and checked in the `FacadeClassLoader.getOrCreateRuntimeClassLoader()` and the `QuarkusTestExtension.doJavaStart()`. I'll still need to look into how to optimize that.

holly-cummins commented on test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestPrepareResult.java:
```diff
@@ -0,0 +1,8 @@
+import io.quarkus.bootstrap.app.AugmentAction;
+import io.quarkus.bootstrap.app.CuratedApplication;
+
+record QuarkusTestPrepareResult(AugmentAction augmentAction, QuarkusTestProfile profileInstance,
```
I don't love the name of this, just because it's so long. I wondered about dropping `Quarkus` since it doesn't add much information (of *course* it's something to do with Quarkus), but then I realised maybe it's supposed to be Quarkus as in `QuarkusTest`. But `QuarkusTest` isn't the right qualifier, because `QuarkusMainTest` also uses it. So I'd vote for sticking with `PrepareResult`. It's not the best name ever, but it's familiar to maintainers, and the signal to noise is slightly higher than `QuarkusTestPrepareResult`.

holly-cummins reviewed (changes requested):
I love to see this, because it's lovely and incremental and exactly what's supposed to happen (but sadly, doesn't always happen) – deliver big change that works but has rough edges, refactor to tidy up rough edges, iterate. 

I spotted one TODO that you may be able to get rid of (or maybe not, but if not then maybe it wants clarifying), and I had a quibble about the naming. I marked it as changes requested but feel free to push back on the naming, because it's so minor and I'm not even sure what the right answer is.

aloubyansky committed:
more

aloubyansky committed:
WIP: AbstractJVMQuarkusTestExtension> do not set TCCL to the SCL if TCCL is FacadeClassLoader (QuarkusMainTest CT issue)

aloubyansky committed:
Added disabled continuous testing test for basic QuarkusMainTest

aloubyansky commented:
I pushed a few more commits. The next 'more' commit will probably be more difficult to review. I moved some code around and removed some.
There was a bit of a mess around the `shutdownTasks`, which I attempted to clean up a bit. It's still not great though.

I tried to add a continuous testing test for `QuarkusMainTest` using out utility methods/framework for CT but it didn't work. This will need some further investigation.

holly-cummins reviewed (approved):
This is all looking excellent! I think this is at a good point to merge and iterate.

aloubyansky commented:
It will be definitely evolving further.
